File: 01-Languages/Python/21-Solutions/solution-06.py
"""
Solution 06 - Generic Stack with Type Hints
See: ../20-Exercises/README.md#exercise-06--generic-stack-with-type-hints-advanced

Run with:
    python solution-06.py

Expected output:
    int stack after pushes: length 3
    Popped: 3
    Peeked (unchanged): 2
    int stack after pop: length 2
    str stack: ['a', 'b']
    Popped from empty stack raised: Stack is empty
"""

from typing import Generic, TypeVar

T = TypeVar("T")


class EmptyStackError(Exception):
    pass


class Stack(Generic[T]):
    """A LIFO stack generic over a single type T (see Lesson 13 for TypeVar/Generic)."""

    def __init__(self) -> None:
        self._items: list[T] = []

    def push(self, item: T) -> None:
        self._items.append(item)

    def pop(self) -> T:
        if self.is_empty():
            raise EmptyStackError("Stack is empty")
        return self._items.pop()

    def peek(self) -> T:
        if self.is_empty():
            raise EmptyStackError("Stack is empty")
        return self._items[-1]

    def is_empty(self) -> bool:
        return len(self._items) == 0

    def __len__(self) -> int:
        return len(self._items)


if __name__ == "__main__":
    # Stack[int] - the [int] is a hint for readers/type-checkers only;
    # the SAME Stack class handles any type at runtime (Lesson 13 explains why).
    int_stack: Stack[int] = Stack()
    int_stack.push(1)
    int_stack.push(2)
    int_stack.push(3)
    print(f"int stack after pushes: length {len(int_stack)}")
    print(f"Popped: {int_stack.pop()}")
    print(f"Peeked (unchanged): {int_stack.peek()}")
    print(f"int stack after pop: length {len(int_stack)}")

    str_stack: Stack[str] = Stack()
    str_stack.push("a")
    str_stack.push("b")
    print(f"str stack: {list(str_stack._items)}")

    empty_stack: Stack[int] = Stack()
    try:
        empty_stack.pop()
    except EmptyStackError as e:
        print(f"Popped from empty stack raised: {e}")

File: 01-Languages/JavaScript/21-Solutions/solution-06.js
// solution-06.js - Stack Class
// See: ../20-Exercises/README.md#exercise-06--stack-class-advanced
//
// Run with:
//   node solution-06.js

class EmptyStackError extends Error {
  constructor(message) {
    super(message);
    this.name = "EmptyStackError";
  }
}

/**
 * A LIFO stack "generic" over a single type T (see Lesson 13: JS has no
 * compile-time generics, so T here is documentation for readers/editors,
 * not something the language enforces - nothing stops mixing types at
 * runtime the way TypeScript's Stack<T> would).
 * @template T
 */
class Stack {
  #items = [];

  /** @param {T} item */
  push(item) {
    this.#items.push(item);
  }

  /** @returns {T} */
  pop() {
    if (this.isEmpty()) throw new EmptyStackError("Stack is empty");
    return this.#items.pop();
  }

  /** @returns {T} */
  peek() {
    if (this.isEmpty()) throw new EmptyStackError("Stack is empty");
    return this.#items[this.#items.length - 1];
  }

  isEmpty() {
    return this.#items.length === 0;
  }

  get size() {
    return this.#items.length;
  }
}

/** @type {Stack<number>} */
const numberStack = new Stack();
numberStack.push(1);
numberStack.push(2);
numberStack.push(3);
console.log(`number stack after pushes: size ${numberStack.size}`);
console.log(`Popped: ${numberStack.pop()}`);
console.log(`Peeked (unchanged): ${numberStack.peek()}`);
console.log(`number stack after pop: size ${numberStack.size}`);

/** @type {Stack<string>} */
const stringStack = new Stack();
stringStack.push("a");
stringStack.push("b");
console.log(`string stack: [${stringStack.peek()}, ...] size ${stringStack.size}`);

const emptyStack = new Stack();
try {
  emptyStack.pop();
} catch (err) {
  console.log(`Popped from empty stack threw: ${err.name}: ${err.message}`);
}
